fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java

```diff
@@ -709,4 +716,207 @@ public Void visitLogicalTableScan(OptExpression optExpression, Void context) {
+        List<ScalarOperator> srcItems = Utils.extractDisjunctive(src);
+        List<ScalarOperator> targetItems = Utils.extractDisjunctive(target);
+        int srcLength = srcItems.size();
+        int targetLength = targetItems.size();
+        for (ScalarOperator item : srcItems) {
+            removeAll(targetItems, item);
+        }
+        if (targetItems.isEmpty() && srcLength == targetLength) {
+            // it is the same, so return true constant
+            return ConstantOperator.createBoolean(true);
+        } else if (!targetItems.isEmpty()) {
+            // the target has more or item, so return src
+            return src;
+        } else {
+            return null;
+        }
+    }
+
+    public static void removeAll(List<ScalarOperator> scalars, ScalarOperator predicate) {
+        Iterator<ScalarOperator> iter = scalars.iterator();
+        while (iter.hasNext()) {
+            ScalarOperator current = iter.next();
+            if (current.equals(predicate)) {
+                iter.remove();
+            }
+        }
+    }
+
+    public static boolean isAllEqualInnerJoin(OptExpression root) {
+        Operator operator = root.getOp();
+        if (!(operator instanceof LogicalOperator)) {
+            return false;
+        }
+        if (operator instanceof LogicalJoinOperator) {
+            LogicalJoinOperator joinOperator = (LogicalJoinOperator) operator;
+            boolean isEqualPredicate = isColumnEqualPredicate(joinOperator.getOnPredicate());
+            if (joinOperator.getJoinType() != JoinOperator.INNER_JOIN || !isEqualPredicate) {
+                return false;
+            }
+        }
+        for (OptExpression child : root.getInputs()) {
+            if (!isAllEqualInnerJoin(child)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public static boolean isColumnEqualPredicate(ScalarOperator predicate) {
+        if (predicate == null) {
+            return false;
+        }
+        if (predicate instanceof BinaryPredicateOperator) {
+            BinaryPredicateOperator binaryPredicate = (BinaryPredicateOperator) predicate;
+            if (binaryPredicate.getBinaryType().isEqual()
+                    && binaryPredicate.getChild(0).isColumnRef()
+                    && binaryPredicate.getChild(1).isColumnRef()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public static Map<ColumnRefOperator, ScalarOperator> getProjectionMap(
+            OptExpression expression, ColumnRefFactory refFactory) {
+        Map<ColumnRefOperator, ScalarOperator> projectionMap;
+        if (expression.getOp().getProjection() != null) {
+            projectionMap = expression.getOp().getProjection().getColumnRefMap();
+        } else {
+            projectionMap = Maps.newHashMap();
+            if (expression.getOp() instanceof LogicalAggregationOperator) {
```

Comment:
why only aggregation, window?

Reply:
only SPJG is supported. window is not supported now.

Reply:
move it to LineageFactory

Reply:
it is not just used in LineageFactory. it is also used in MaterializedViewRewriter. So I put it here.